File: pages/ListingProductPage.js
const { By, until } = require("selenium-webdriver");

class ListingProductPage {
  constructor(driver) {
    this.driver = driver;
  }

  listingProductBtn = By.xpath("//a[contains(@id,'listing_product')]");

  // Tìm thẻ chứa text, sau đó lùi lên trên (ancestor) để tìm thẻ div bao ngoài có class chứa '-control'
  channelDropdown = By.xpath(
    "//div[text()='Chọn kênh bán hàng']/ancestor::div[contains(@class, '-control')]",
  );

  // Thay đổi: dùng contains(., ...) thay vì contains(text(), ...)
  // Đồng thời dùng class chứa '-option' để chỉ định chính xác các hàng trong menu
  channelItem = (channelName) =>
    By.xpath(
      `//div[contains(@class, '-menu')]//div[contains(@class, '-option') and contains(., '${channelName}')]`,
    );

  continueBtn = By.xpath("//button[contains(.,'Tiếp theo')]");

  sellProductBtn = By.xpath("//button[contains(.,'Đăng bán')]");

  async accessListingProduct() {
    await this.driver.sleep(2000);
    const originalTab = await this.driver.getWindowHandle();
    const listingBtn = await this.driver.wait(
      until.elementLocated(this.listingProductBtn),
      10000,
    );
    await this.driver.wait(until.elementIsVisible(listingBtn), 10000);
    await listingBtn.click();

    await this.driver.wait(async () => {
      const handles = await this.driver.getAllWindowHandles();
      return handles.length > 1;
    }, 10000);

    const allTabs = await this.driver.getAllWindowHandles();
    const newTab = allTabs.find((tab) => tab !== originalTab);
    await this.driver.switchTo().window(newTab);
    console.log("Switched to listing tab");
  }

  async selectSalesChannel() {
    // 1. Chờ element xuất hiện trong DOM (tối đa 10 giây)
    const dropdown = await this.driver.wait(
      until.elementLocated(this.channelDropdown),
      10000,
    );

    // 2. Chờ element hiển thị trên màn hình và có thể tương tác
    await this.driver.wait(until.elementIsVisible(dropdown), 10000);

    // 3. Cuộn chuột đến element (đề phòng dropdown nằm ngoài tầm mắt)
    await this.driver.executeScript(
      "arguments[0].scrollIntoView(true);",
      dropdown,
    );

    // 4. Thực hiện click
    await dropdown.click();
    console.log("Clicked sales channel dropdown");
  }

  async selectChannelOption(channelName) {
    // 1. Định vị item cần chọn (bây giờ XPath đã quét được toàn bộ text bao gồm cả sau icon)
    const itemLocator = this.channelItem(channelName);

    // 2. Chờ item đó xuất hiện trong DOM
    const option = await this.driver.wait(
      until.elementLocated(itemLocator),
      10000,
    );

    // 3. Cuộn đến option đó (đề phòng danh sách dài bị khuất)
    await this.driver.executeScript(
      "arguments[0].scrollIntoView(true);",
      option,
    );

    // 4. Click thẳng bằng JavaScript để né việc click lệch vào cái Icon ngôi nhà
    await this.driver.executeScript("arguments[0].click();", option);
    console.log(`Selected channel: ${channelName} via JS successfully`);
  }

  async clickContinue() {
    const continueButton = await this.driver.wait(
      until.elementLocated(this.continueBtn),
      10000,
    );
    await this.driver.wait(until.elementIsVisible(continueButton), 10000);
    await continueButton.click();
  }

  async clickSellProduct() {
    const sellButton = await this.driver.wait(
      until.elementLocated(this.sellProductBtn),
      10000,
    );
    await this.driver.wait(until.elementIsVisible(sellButton), 10000);
    await this.driver.sleep(2000); // thêm sleep để đảm bảo mọi thứ đã sẵn sàng trước khi click
    await sellButton.click();
    await this.driver.sleep(2000); // thêm sleep để chờ phản hồi sau khi click
  }
}

module.exports = ListingProductPage;

File: src/test/java/com/example/seleniumtestng/pages/ListingProductPage.java
package com.example.seleniumtestng.pages;

import java.util.List;
import java.util.Set;
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class ListingProductPage extends BasePage {
    private final By listingProductBtn = By.xpath("//a[contains(@id,'listing_product')]");
    private final By channelDropdown = By.xpath("//div[text()='Chọn kênh bán hàng']/ancestor::div[contains(@class,'-control')]");
    private final By continueBtn = By.xpath("//button[contains(.,'Tiếp theo')]");
    private final By createListingStep = By.xpath("//*[contains(normalize-space(.),'Tạo sản phẩm')]");
    private final By loadingText = By.xpath("//*[normalize-space()='Đang tải']");
    private final By sellProductBtn = By.xpath("//button[contains(.,'Đăng bán')]");
    private final By sellProductButtonLike = By.xpath("//*[self::button or @role='button' or contains(@class,'btn')][contains(normalize-space(.),'Đăng bán')]");
    private WebElement lastSalesChannelDropdown;

    public ListingProductPage(WebDriver driver) {
        super(driver);
    }

    public void accessListingProduct() {
        sleep(2000);
        Set<String> originalHandles = driver.getWindowHandles();
        WebElement listingButton = visible(listingProductBtn);
        try {
            listingButton.click();
        } catch (RuntimeException e) {
            jsClick(listingButton);
        }

        try {
            shortWait(10000).until(driver -> driver.getWindowHandles().size() > originalHandles.size());
        } catch (RuntimeException ignored) {
        }

        for (String handle : driver.getWindowHandles()) {
            if (!originalHandles.contains(handle)) {
                driver.switchTo().window(handle);
                System.out.println("Switched to new listing tab");
                return;
            }
        }
        System.out.println("No new tab opened for listing; staying in current tab");
    }

    public void selectSalesChannel() {
        lastSalesChannelDropdown = visibleFirst(
                channelDropdown,
                By.xpath("//div[contains(normalize-space(.),'Chọn kênh bán hàng')]/ancestor::div[contains(@class,'-control')][1]"),
                By.xpath("//input[contains(@id,'react-select')]/ancestor::div[contains(@class,'-control')][1]"));
        try {
            lastSalesChannelDropdown.click();
        } catch (RuntimeException e) {
            jsClick(lastSalesChannelDropdown);
        }
        sleep(500);
        System.out.println("Clicked sales channel dropdown");
    }

    public void selectChannelOption(String channelName) {
        By optionLocator = By.xpath("//div[contains(@class,'-menu')]//div[contains(@class,'-option') and contains(.,'" + channelName + "')]");
        WebElement option = findVisibleOption(optionLocator, 3000);
        if (option == null) {
            typeChannelName(channelName);
            option = findVisibleOption(optionLocator, 5000);
        }

        if (option != null) {
            jsClick(option);
        } else {
            driver.switchTo().activeElement().sendKeys(Keys.ENTER);
        }
        System.out.println("Selected channel: " + channelName);
    }

    public void clickContinue() {
        WebElement continueButton = clickable(continueBtn);
        try {
            continueButton.click();
        } catch (RuntimeException e) {
            jsClick(continueButton);
        }
        visible(createListingStep);
        waitForListingForm();
    }

    public void clickSellProduct() {
        WebElement sellButton = waitForSellButton(60000);
        if (sellButton == null) {
            driver.navigate().refresh();
            visible(createListingStep);
            sellButton = waitForSellButton(60000);
        }
        if (sellButton == null) {
            throw new IllegalStateException("Không tìm thấy nút Đăng bán trên trang listing. URL hiện tại: " + driver.getCurrentUrl());
        }
        try {
            sellButton.click();
        } catch (RuntimeException e) {
            jsClick(sellButton);
        }
        sleep(2000);
    }

    public void closeExtraWindows(String mainWindowHandle) {
        List<String> handles = List.copyOf(driver.getWindowHandles());
        for (String handle : handles) {
            if (!handle.equals(mainWindowHandle)) {
                driver.switchTo().window(handle);
                driver.close();
            }
        }
        driver.switchTo().window(mainWindowHandle);
    }

    private WebElement visibleFirst(By... locators) {
        return wait.until(driver -> {
            for (By locator : locators) {
                for (WebElement element : driver.findElements(locator)) {
                    try {
                        if (element.isDisplayed()) {
                            return element;
                        }
                    } catch (RuntimeException ignored) {
                    }
                }
            }
            return null;
        });
    }

    private WebElement findVisibleOption(By locator, long timeoutMillis) {
        try {
            return shortWait(timeoutMillis).until(driver -> {
                for (WebElement element : driver.findElements(locator)) {
                    try {
                        if (element.isDisplayed()) {
                            return element;
                        }
                    } catch (RuntimeException ignored) {
                    }
                }
                return null;
            });
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    private void waitForListingForm() {
        waitForSellButton(60000);
    }

    private WebElement waitForSellButton(long timeoutMillis) {
        try {
            return shortWait(timeoutMillis).until(driver -> {
                WebElement button = firstDisplayed(sellProductBtn, sellProductButtonLike);
                if (button != null && button.isEnabled()) {
                    return button;
                }
                if (isLoadingVisible()) {
                    return null;
                }
                return button;
            });
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    private WebElement firstDisplayed(By... locators) {
        for (By locator : locators) {
            for (WebElement element : driver.findElements(locator)) {
                try {
                    if (element.isDisplayed()) {
                        return element;
                    }
                } catch (RuntimeException ignored) {
                }
            }
        }
        return null;
    }

    private boolean isLoadingVisible() {
        for (WebElement element : driver.findElements(loadingText)) {
            try {
                if (element.isDisplayed()) {
                    return true;
                }
            } catch (RuntimeException ignored) {
            }
        }
        return false;
    }

    private void typeChannelName(String channelName) {
        WebElement input = null;
        if (lastSalesChannelDropdown != null) {
            List<WebElement> inputs = lastSalesChannelDropdown.findElements(By.cssSelector("input"));
            if (!inputs.isEmpty()) {
                input = inputs.get(0);
            }
        }
        if (input == null) {
            input = driver.switchTo().activeElement();
        }
        input.sendKeys(channelName);
        sleep(500);
    }

    private void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while waiting", e);
        }
    }
}
